diff --git a/kernel/src/arch/arm/mod.rs b/kernel/src/arch/arm/mod.rs
index b617fa3..55c2cc1 100644
--- a/kernel/src/arch/arm/mod.rs
+++ b/kernel/src/arch/arm/mod.rs
@@ -33,9 +33,9 @@ use scheduler::ContextSwitchHookHolder;
 
 pub const EXCEPTION_LR: usize = 0xFFFFFFFD;
 // See https://developer.arm.com/documentation/100235/0100/The-Cortex-M33-Processor/Programmer-s-model/Core-registers/CONTROL-register.
-#[cfg(not(target_abi = "eabihf"))]
+#[cfg(not(has_fpu))]
 pub const CONTROL: usize = 0b10;
-#[cfg(target_abi = "eabihf")]
+#[cfg(has_fpu)]
 pub const CONTROL: usize = 0b110;
 pub const THUMB_MODE: usize = 0x01000000;
 pub const NR_SWITCH: usize = !0;
@@ -118,7 +118,7 @@ pub extern "C" fn start_schedule(cont: extern "C" fn() -> !) {
     unsafe { reset_msp_and_start_schedule(&mut __sys_stack_end as *mut u8, cont) }
 }
 
-#[cfg(not(target_abi = "eabihf"))]
+#[cfg(not(has_fpu))]
 #[repr(C, align(8))]
 #[derive(Default, Debug, Copy, Clone)]
 pub struct Context {
@@ -143,7 +143,7 @@ pub struct Context {
     pub xpsr: usize,
 }
 
-#[cfg(target_abi = "eabihf")]
+#[cfg(has_fpu)]
 #[repr(C, align(8))]
 #[derive(Default, Debug, Copy, Clone)]
 pub struct Context {
@@ -202,7 +202,7 @@ pub struct Context {
     pub vpr: usize,
 }
 
-#[cfg(not(target_abi = "eabihf"))]
+#[cfg(not(has_fpu))]
 #[repr(C, align(8))]
 #[derive(Default)]
 pub struct IsrContext {
@@ -217,7 +217,7 @@ pub struct IsrContext {
 }
 
 // See https://developer.arm.com/documentation/107706/0100/Exceptions-and-interrupts-overview/Stack-frames.
-#[cfg(target_abi = "eabihf")]
+#[cfg(has_fpu)]
 #[repr(C, align(8))]
 #[derive(Default)]
 pub struct IsrContext {
@@ -260,7 +260,7 @@ impl fmt::Debug for IsrContext {
         write!(f, "lr: 0x{:x} ", self.lr)?;
         write!(f, "pc: 0x{:x} ", self.pc)?;
         write!(f, "xpsr: 0x{:x} ", self.xpsr)?;
-        #[cfg(target_abi = "eabihf")]
+        #[cfg(has_fpu)]
         {
             write!(f, "fpscr: 0x{:x} ", self.fpscr)?;
             write!(f, "vpr: 0x{:x} ", self.vpr)?;
@@ -272,7 +272,7 @@ impl fmt::Debug for IsrContext {
 
 // FIXME: We need to pass a scratch register to perform saving.
 // Use r12 as scratch register now.
-#[cfg(not(target_abi = "eabihf"))]
+#[cfg(not(has_fpu))]
 macro_rules! store_callee_saved_regs {
     () => {
         "
@@ -282,7 +282,7 @@ macro_rules! store_callee_saved_regs {
     };
 }
 
-#[cfg(not(target_abi = "eabihf"))]
+#[cfg(not(has_fpu))]
 macro_rules! load_callee_saved_regs {
     () => {
         "
@@ -292,7 +292,7 @@ macro_rules! load_callee_saved_regs {
     };
 }
 
-#[cfg(target_abi = "eabihf")]
+#[cfg(has_fpu)]
 macro_rules! store_callee_saved_regs {
     () => {
         "
@@ -303,7 +303,7 @@ macro_rules! store_callee_saved_regs {
     };
 }
 
-#[cfg(target_abi = "eabihf")]
+#[cfg(has_fpu)]
 macro_rules! load_callee_saved_regs {
     () => {
         "
@@ -480,7 +480,7 @@ impl Context {
         self
     }
 
-    #[cfg(not(target_abi = "eabihf"))]
+    #[cfg(not(has_fpu))]
     #[inline]
     pub fn init(&mut self) -> &mut Self {
         self.xpsr = THUMB_MODE;
@@ -488,7 +488,7 @@ impl Context {
     }
 
     // See https://developer.arm.com/documentation/100235/0004/the-cortex-m33-peripherals/floating-point-unit/floating-point-status-control-register.
-    #[cfg(target_abi = "eabihf")]
+    #[cfg(has_fpu)]
     #[inline]
     pub fn init(&mut self) -> &mut Self {
         self.xpsr = THUMB_MODE;
@@ -655,7 +655,7 @@ mod tests {
     // See https://developer.arm.com/documentation/107706/0100/Exceptions-and-interrupts-overview/Stack-frames.
     #[test]
     fn test_abi() {
-        #[cfg(target_abi = "eabihf")]
+        #[cfg(has_fpu)]
         {
             assert_eq!(
                 core::mem::size_of::<IsrContext>(),
@@ -666,7 +666,7 @@ mod tests {
                 core::mem::size_of::<IsrContext>() + 8 * 4 + 16 * 4
             );
         }
-        #[cfg(not(target_abi = "eabihf"))]
+        #[cfg(not(has_fpu))]
         {
             assert_eq!(
                 core::mem::size_of::<IsrContext>(),
diff --git a/kernel/src/boards/qemu_mps3_an547/mod.rs b/kernel/src/boards/qemu_mps3_an547/mod.rs
index bedd0d4..819722d 100644
--- a/kernel/src/boards/qemu_mps3_an547/mod.rs
+++ b/kernel/src/boards/qemu_mps3_an547/mod.rs
@@ -87,7 +87,10 @@ const HZ: usize = SYSTEM_CORE_CLOCK as usize;
 pub type ClockImpl = systick::SysTickClock<TICKS_PS, HZ>;
 
 pub(crate) fn init() {
-    unsafe { enable_fpu() };
+    #[cfg(has_fpu)]
+    unsafe {
+        enable_fpu()
+    };
 
     unsafe {
         copy_data();
